_scripts/git-precommit-hook.sh

```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+# 1. Get the list of all modified and dependent packages.
+# Cleans up yarn's output by removing everything before the package name, the first and the last line.
+PACKAGES_MODIFIED=`yarn workspaces list --since=HEAD^ -R | cut -d/ -f2 | sed '1d' | sed '$d'`
+
+# 2. If no packages were modified, return early; otherwise we'd build all packages.
+if [[ $PACKAGES_MODIFIED == "" ]]; then
+    echo "no modified packages found"
+    exit 0
+fi
+
+echo "modified packages found"
+
+# 3. Build all back-end TS packages in the list
+ORIGINAL_IFS=$IFS
+IFS=$'\n'
+# We don't need to worry about front-end packages that use TS, since they must build successfully for CI to pass.
+BACKEND_PACKAGES=( "fxa-auth-server" "fxa-event-broker" "fxa-graphql-api" "fxa-support-panel" "fxa-admin-server" )
+INCLUDE_ARGS=''
+for package_modified in $PACKAGES_MODIFIED; do
+  if [[ " ${BACKEND_PACKAGES[@]} " =~ " ${package_modified} " ]]; then
+      INCLUDE_ARGS="$INCLUDE_ARGS --include ${package_modified}"
+  fi
+done
+IFS=$ORIGINAL_IFS
+INCLUDE_ARGS=`echo "$INCLUDE_ARGS" | xargs` # trim whitespace at beginning of string
+echo "building all modified and dependent backend TS packages..."
+# We don't need to write files to disk, so use --noEmit here instead of --build for speed.
+if ! `yarn workspaces foreach --verbose ${INCLUDE_ARGS} run tsc --noEmit > artifacts/build-affected-backend-packages.log`;
```

Comment:
Q: Should we add `--parallel`?

Reply:
Also, not sure if I'm missing something, but I don't think this is working for me locally.

I'm trying to hardcode the command in my Terminal; but it's very fast and suspiciously unverbose:
```sh
yarn --version # 3.1.1
yarn workspaces foreach --verbose --include fxa-event-broker --include fxa-admin-server --include fxa-graphql-api --include fxa-auth-server -p run tsc --noEmit
➤ YN0000: Done in 0s
```

But, if I change <kbd>run tsc --noEmit</kbd> to <kbd>run build</kbd> then it seems to work and run the scripts. So, looks like I can run scripts from package.json's, but not arbitrary commands:
```sh
yarn workspaces foreach --verbose --include fxa-event-broker --include fxa-admin-server --include fxa-graphql-api --include fxa-auth-server -p run build
➤ YN0000: [fxa-graphql-api]: Process exited (exit code 0), completed in 6s 339ms
➤ YN0000: [fxa-event-broker]: Process exited (exit code 0), completed in 7s 188ms
➤ YN0000: [fxa-admin-server]: Found 4 error(s).
➤ YN0000: [fxa-admin-server]: (.....a few errors.....)
➤ YN0000: [fxa-admin-server]: Process exited (exit code 1), completed in 14s 781ms
➤ YN0000: [fxa-auth-server]: (node:61006) [DEP0147] DeprecationWarning: In future versions of Node.js, fs.rmdir(path, { recursive: true }) will be removed. Use fs.rm(path, { recursive: true }) instead
➤ YN0000: [fxa-auth-server]: (Use `node --trace-deprecation ...` to show where the warning was created)
➤ YN0000: [fxa-auth-server]: Process exited (exit code 0), completed in 20s 376ms
➤ YN0000: Done in 20s 386ms
```

Reply:
Ah thanks; you're right. My husband was rushing me for dinner. I'm not trying to blame him, but it's definitely his fault. I'll update the script.

Reply:
> Q: Should we add `--parallel`?

I don't think so, because some packages depend on other packages, so we want to compile those independent packages first (e.g. `fxa-shared` should be compiled before `fxa-auth-server`, since the latter depends on it).

Although... I'm not exactly sure how `--noEmit` works though, since for `--build`, we'd get a physical built file that can be included in the next package's compilation. 🤔  I'll see if I can find out.

Reply:
I just ran it with `--parallel`, and it looks like it still ran and completed `fxa-shared` first before parallelizing the other packages. So seems like it'd work? I got 24s down from 37s.

Unfortunately, despite [yarn's docs](https://yarnpkg.com/cli/workspaces/foreach) stating the opposite, the output seems to be interlaced[1]:
> If -p,--parallel and -i,--interlaced are both set, Yarn will print the lines from the output as it receives them. If -i,--interlaced wasn't set, it would instead buffer the output from each process and print the resulting buffers only after their source processes have exited.

[1]:
```
➤ YN0000: [fxa-shared]: Process started
➤ YN0000: [fxa-shared]: Process exited (exit code 0), completed in 9s 418ms
➤ YN0000: [fxa-admin-server]: Process started
➤ YN0000: [fxa-auth-server]: Process started
➤ YN0000: [fxa-event-broker]: Process started
➤ YN0000: [fxa-support-panel]: Process started
➤ YN0000: [fxa-support-panel]: Process exited (exit code 0), completed in 3s 18ms
➤ YN0000: [fxa-event-broker]: Process exited (exit code 0), completed in 6s 397ms
➤ YN0000: [fxa-admin-server]: Process exited (exit code 0), completed in 8s 802ms
➤ YN0000: [fxa-auth-server]: Process exited (exit code 0), completed in 10s 734ms
➤ YN0000: [fxa-graphql-api]: Process started
➤ YN0000: [fxa-graphql-api]: Process exited (exit code 0), completed in 3s 886ms
➤ YN0000: Done in 24s 52ms
```